New ability to work in offline mode, clearing unnecessary

The-going opened the pull request:
There are currently two questionable variables.
    They are the keys in the structures:
  ` if [[ $FORCE_CHECKOUT == yes ]]; then` in the function `fetch_from_repo()`
    which is set by default by the script 'main.sh'
  ```
 [[ -z $FORCE_CHECKOUT ]] && FORCE_CHECKOUT=yes
    ....
```
    
    Next in the script 'main.sh'
 ```

    # ignore updates help on building all images - for internal purposes
    # fetch_from_repo <url> <dir> <ref> <subdir_flag>
    if [[ $IGNORE_UPDATES != yes ]]; then
        display_alert "Downloading sources" "" "info"
        fetch_from_repo "$BOOTSOURCE" "$BOOTDIR" "$BOOTBRANCH" "yes"
        fetch_from_repo "$KERNELSOURCE" "$KERNELDIR" "$KERNELBRANCH" "yes"
    
```
    What does this key imply?
    It only makes sense in one case if we control the build from an external
    script and clearing repositories somewhere in it.
    
    We should always clear the sources before starting the build.
    For this reason, these two keys in these places do not make any sense.
    You can just delete them.
    It is much more important to be able to work offline.
    This is useful when building recursively for multiple platforms,
    and some resource on the Internet is frozen.
    The process is stopped and waiting.
    I think everyone understands what we are talking about.
    To do this, I entered a new variable 'OFFLINE_WORK'
    
    When 'OFFLINE_WORK=yes' we work offline and
    just return the sources to their original state.
    
    Fix issue #1952

The-going commented:
Currently, the 'fetch_from_repo' function performs two actions. 
1) Checks the sources and updates if they are old. 
2) Returns git to its original state.
 It was logical to move the IGNORE_UPDATES action inside the 'fetch_from_repo' function.
Now this is the 'OFFLINE_WORK' name for the bypass (1)

The-going commented:
With the latest updates, my code stopped working correctly. I'm checking again today.

igorpecovnik commented:
You mean with this?
https://github.com/armbian/build/commit/f8496ad91485ec1fb81cc06ad974579e4989755d

zhangn1985 commented:
please also remove unrelated changes.

The-going commented:
Found it! Updates are not to blame.
The VM disk has run out of free space. I created and mounted an additional disk. Created the `sources` folder there and a symbolic link from `build/cache`. Now there are two ways to get to `sources`.
Now this comparison is always performed. [#L248](https://github.com/The-going/armbian-build/blob/master/lib/general.sh#L248)
The `$(pwd)` command returns the real directory.
`$(git rev-parse --show-toplevel)` - path of the symbolic link.
Checking this condition is unnecessary here.
We just new created an empty folder in this path and\or made a transition to an existing one.
All we need to do is check whether the git folder is a repository.
And the remote URL matches the target URL.
  [#L248](https://github.com/The-going/armbian-build/blob/master/lib/general.sh#L248) and   [#L241](https://github.com/The-going/armbian-build/blob/master/lib/general.sh#L241) I removed both from the comparison conditions, and everything worked.
Tomorrow I will twist this situation again.

The-going commented:
* Added an exit when you can't create a directory
The case when the broken link is in the path or there are no write rights. The disk was mounted for read only.
 * Fixed checking folder as git repository and target URL
The previous code worked incorrectly when the folder is not git it returned the nearest git when moving to '/'
 * Removed non topic related changes from the commit

zhangn1985 commented:
@The-going do you try clean build? clone armbian/build + your PR in another folder and do clean build.

The-going commented:
> @The-going do you try clean build? clone armbian/build + your PR in another folder and do clean build.

Yes. I scrolled through all the options that turned out to be a rake for me and those that could potentially happen. 
But for the sake of cleanliness, I'll do it one more time. I will divide PR into two fixations.
When I test the original armbian/build, I will surround [this line](https://github.com/armbian/build/blob/master/lib/general.sh#L262) with debugging code and post the log here.

The-going commented:
TEST (1) - broken link in the path ~/build/cache/sources
broken link in the path ~/build/cache/sources
the entire file structure of the build folder is killed.
The branch switched to FETCH_HEAD and despite obvious errors, the script continued to update the already empty folder /build/cache/sources. stopped manually.
`cd $HOME`
`ls -l build/`
`drwxrwxr-x   2 root root   4096 мая 25 20:51 api`
`drwxrwxr-x  15 root root   4096 мая 25 20:51 arch`
`drwxrwxr-x 190 root root   4096 мая 25 20:51 board`
`drwxrwxr-x   3 root root   4096 мая 25 20:51 cache`
`drwxrwxr-x   7 root root   4096 мая 25 20:51 cmd`
`drwxrwxr-x   5 root root   4096 мая 25 20:51 common`
`-rw-rw-r--   1 root root   2196 мая 25 20:51 config.mk`
`drwxrwxr-x   2 root root  69632 мая 25 20:51 configs`
`drwxrwxr-x   2 root root   4096 мая 25 20:51 disk`
`drwxrwxr-x  19 root root   4096 мая 25 20:51 doc`
`drwxrwxr-x  65 root root   4096 мая 25 20:51 drivers`
`drwxrwxr-x   2 root root   4096 мая 25 20:51 dts`
`drwxrwxr-x   2 root root   4096 мая 25 20:51 env`
`drwxrwxr-x   4 root root   4096 мая 25 20:51 examples`
`drwxrwxr-x  13 root root   4096 мая 25 20:51 fs`
`drwxrwxr-x  30 root root  20480 мая 25 20:51 include`
`-rw-rw-r--   1 root root   1863 мая 25 20:51 Kbuild`
`-rw-rw-r--   1 root root  22173 мая 25 20:51 Kconfig`
`drwxrwxr-x  20 root root   4096 мая 25 20:51 lib`
`drwxrwxr-x   2 root root   4096 мая 25 20:51 Licenses`
`-rw-rw-r--   1 root root  26256 мая 25 20:51 MAINTAINERS`
`-rw-rw-r--   1 root root  74218 мая 25 20:51 Makefile`
.......

The-going commented:
I was lucky. The entire cache on the other drives. Only symbolic links were removed.
Restored the build directory structure.
I cleared the uboot repository, and the rest is complete.

output.log >>>>
Displaying message: Checking git sources u-boot v2020.04 info
Displaying message: DEBUG $current_url []
Displaying message: DEBUG $url [git://git.denx.de/u-boot.git]
Displaying message: DEBUG $(git rev-parse --is-inside-work-tree 2>/dev/null) []
Displaying message: DEBUG $(git rev-parse --show-toplevel) []
Displaying message: DEBUG $(pwd) [/home/leo/build/cache/sources/u-boot/v2020.04]
Displaying message: Creating local copy
Displaying message: Fetching updates
Displaying message: Checking out
Displaying message: Checking git sources linux-mainline orange-pi-5.4 info
Displaying message: DEBUG $current_url [https://github.com/megous/linux]
Displaying message: DEBUG $url [https://github.com/megous/linux]
Displaying message: DEBUG $(git rev-parse --is-inside-work-tree 2>/dev/null) [true]
Displaying message: DEBUG $(git rev-parse --show-toplevel) [/home/leo/build_cache/sources/linux-mainline/orange-pi-5.4]
Displaying message: DEBUG $(pwd) [/home/leo/build/cache/sources/linux-mainline/orange-pi-5.4]
Displaying message: Creating local copy
fatal: an external origin repository already exists
Displaying message:  Cleaning ....  2668 files
Displaying message: Checking git sources arm-trusted-firmware-sunxi-mainline master info
Displaying message: DEBUG $current_url [https://github.com/ARM-software/arm-trusted-firmware]
Displaying message: DEBUG $url [https://github.com/ARM-software/arm-trusted-firmware]
Displaying message: DEBUG $(git rev-parse --is-inside-work-tree 2>/dev/null) [true]
Displaying message: DEBUG $(git rev-parse --show-toplevel) [/home/leo/build_cache/sources/arm-trusted-firmware-sunxi-mainline/master]
Displaying message: DEBUG $(pwd) [/home/leo/build/cache/sources/arm-trusted-firmware-sunxi-mainline/master]
Displaying message: Creating local copy
fatal: an external origin repository already exists
Displaying message:  Cleaning ....  1 files

The-going commented:
[DEBUG code](https://github.com/The-going/armbian-build/blob/testmaster/lib/general.sh#L263)

The-going commented:
I divide commit into two parts.
To be continued.

The-going commented:
[The first commit in this PR](https://github.com/armbian/build/pull/1998)

The-going commented:
This PR has grown old

Full version [here](https://github.com/The-going/armbian-build/tree/offlinework)
Now you can collect packages without accessing the Internet

igorpecovnik commented on lib/general.sh:
```diff
@@ -204,8 +204,16 @@ fetch_from_repo()
 	local ref_subdir=$4
 
+	# The 'offline' variable must always be set to 'true' or 'false'
+	if [ "$OFFLINE_WORK" == "yes" ]; then
```
OFFLINE_WORK
Also add here https://github.com/armbian/build/blob/master/lib/build-all-ng.sh#L60 and 
https://github.com/armbian/documentation/blob/master/docs/Developer-Guide_Build-Options.md

The-going replied:
OK!

The-going replied:
[.ignore_changes](https://github.com/armbian/documentation/blame/master/docs/Developer-Guide_Build-Options.md#L79)
This strange entity made me think.
It is only [present here](https://github.com/armbian/build/blob/master/compile.sh#L44)
But this code in my build system never works, because the [systemd-detect-virt](https://github.com/armbian/build/blob/master/compile.sh#L77) command always returns "kvm"
For me, the lack of this functionality is a great boon, updating the build system manually is what I most want.
What should I do in this script?

zhangn1985 replied:
do you run your build on cloud?

The-going replied:
If the build system is in the cloud, there are no problems with the Internet, and there are no problems with the build speed. In this case, I wouldn't be doing this job.
I have openSUSE working on my host and use qemu-kvm and virt-manager as a VM.
